### ISSUE TITLE
Should hook reactors be able to abort a hit?

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#640]
This came up in https://github.com/WordPoints/dynamic-points/issues/9, and it was decided that it was a question that needed to be answered in core.

Basically, there are times when a reactor is unable to respond to a fire as it normally would. For example, the target arg may be an "optional" relationship (like a comment author). When the value of the arg is not set, the reactor won't be able to target it.

However, the API is designed such that reactors are only called into play after the hit has already been recorded, so they have no ability to abort it, even if they abort its effects.

This affects the behavior of periods, rates/limits (once implemented, #472, #471), repeatability, and maybe reversals. The resulting behavior will likely be unexpected to some users in some cases, though possibly not to all users in all cases—which is why there is a dilemma here.

Probably there is an elegant way of satisfying both a sane default the option for settings to override it, without destroying the distinction between reactors and extensions within the API.

**Comment by JDGrimes** [comment]
This came up in https://github.com/WordPoints/wordpoints/issues/762#issuecomment-385062439, in regard to checking whether the target was valid during validation (specifically, so that we wouldn't end up recording hits for guests).

Looking back over https://github.com/WordPoints/dynamic-points/issues/9#issuecomment-284421728, this paragraph stands out:

>The real underlying question though is what the expectations of the user would be in regard to repeatability and rates and limits. If a user only wants to award points for five matching comments, they probably don't want to have guests end up leaving some of the matching comments and thus registered users not end up getting those points at all. The idea of a hit is that it will have the intended affect—the reactor will react. Which in this case it can't.

The key word here is "probably" in the second sentence. There is the possibility that the user would want guest comments to be counted (and in the future we might make guest points a possibility). But if the reactor validates the target, then we no longer have the possibility of giving the user this choice.

**Comment by JDGrimes** [comment]
I think though, that in this particular scenario, the main reason to hit for guests is if we want to have the possibility of actually awarding points to them. But when that possibility is introduced, it would be up the reactor to change its code to accommodate that.

So setting aside the idea that the action might actually have the intended effect, that is, considering the case where the reactor will not actually react, what then?

The fact remains that the event has indeed still taken place. And all of the extension requirements were met. In that sense, a "hit" has occurred. The question would come down to whether a hit ought to indicate that a matching fire occurred, or that the reactor actually completed the action.

Given that we didn't provide for reactor result status to be returned and have any effect, I think that we've already answered this question.

As such, allowing the reactor to abort a fire on an arbitrary basis would run counter to the entire design of the system.

But there is still the argument to be made that in some cases we aren't trying to abort a fire from the reactor on an arbitrary basis, but based on an unmet validation point. A reactor can indeed abort based on failed validation, and the question is where do we draw the line between validation and a condition. This is actually a reactor-specific question, but there may be a general principle that applies across the board.

Currently, reactors abort during validation if their settings aren't met. The parent reactor class specifically checks that the target is set and that the arg actually is present in the hierarchy. This seems like straightforward reaction settings validation that is not tied to the current fire and context. However, the existence of optional args or relationships or attributes leads to the place where even this is not necessarily guaranteed. So in that sense, our current code may already be blurring the lines between reaction validation and fire validation.

**Comment by JDGrimes** [comment]
That said, we really don't have optional relationships or args as part of the design of the API at this point.  It is more of a future possibility.

So in general, I think that we can say that a reaction should stick with validation of the reaction settings, not validation of the fire against those settings. Any other behavior would have to be considered on a per-reactor basis. Other behavior being implied requirements regarding some of the settings as they relate to the context of the fire, not arbitrary requirements having nothing to do with the reactor's function itself.

This would seem to answer the basic question at the start of this ticket. Although further consideration of the exact behavior of the points reactor specifically is still called for, that is really an independent question.

**Comment by JDGrimes** [comment]
How would a reactor go about doing this kind of validation if it wanted to, though?

We have only provided for reaction settings validation in the API's design, so any further validation by the reactor would have to be done independently of the reaction settings validation. Specifically, we have no way of knowing for sure during reaction settings validation whether we are checking the settings in context of a fire or not.

The problem of course, is that once the reactor has the opportunity to abort relative to the context, that feature is open for abuse.

We haven't yet come to a sure conclusion on whether this is even a good idea within the limited case of the points reactor yet, though. So that is what we'd need to figure out before it would make sense to proceed here.

**Comment by JDGrimes** [comment]
Consider a limit that allows only 5 different users to be awarded. Each guest will not count as a different user at present, because they all simply have the same ID (or lack thereof). Therefore, 5 guests triggering a hit will not cause the limit to be increased; only the first guest will be awarded. A change in how guest users are handled across the entire API would be necessary to accommodate a different result.

The same goes for rates. And as for repeatability and reversals, guests would all be in the same repeat/reversal pool, not independent.

As such, the API does not really handle guest users in these other places where these hits would actually come into play either. Of course, since no points are awarded to guests, the reversals and repeatability are moot in regard to them. The rates and limits, however, would affect registered users, but would not treat guests as individuals, but as a collective. There may be rare times when this behavior is desired, but it is certainly not a normal use-case.

What this means is that the behavior currently made possible by tracking such non-reacted hits is not useful. So tracking these hits is a waste of space in the DB.

As such, it would be better to abort the fire early when the reactor knows that it cannot actually handle the fire in question.

In short, guest users are not handled by any part of the API's present design, yet hits are still recorded. When the API does handle guests, these hits will be useless, for the most part, because they will not have any differentiating information. So as far as the points reactor goes, it makes sense for it to be able to abort fires that take place in an invalid context. Just as, indeed, these extensions would have to do to maintain sane behavior (at least at this time).

**Comment by JDGrimes** [comment]
However, what is necessary at this time is not a general ability to check the context, but only a specific need to ensure that the target actually exists. As such, it might be possible to build this check into the base reactor rather than introducing a new interface and method that each reactor would have to implement. This would limit the potential for abuse.

The question is, do we want this check to be hard-coded into the base reactor or some other part of the API, or do we want to make it possible for a reactor to accept hits without a valid target? On the flip-side, perhaps a reactor could require more of the target than just the ID, like some other attribute (for example, the user email if the reactor sends notifications, etc.).

So I think it makes sense to give reactors the opportunity to specifically validate the target.
